executor/union_scan.go

```diff
@@ -0,0 +1,243 @@
+package executor
+
+import (
+	"sort"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/evaluator"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/util/types"
+)
+
+type dirtyDB struct {
+	tables map[int64]*dirtyTable
+}
+
+func (udb *dirtyDB) AddRow(tid, handle int64, row []types.Datum) {
+	dt := udb.getDirtyTable(tid)
+	for i := range row {
+		if row[i].Kind() == types.KindString {
+			row[i].SetBytes(row[i].GetBytes())
+		}
+	}
+	dt.addedRows[handle] = row
+}
+
+func (udb *dirtyDB) DeleteRow(tid int64, handle int64) {
+	dt := udb.getDirtyTable(tid)
+	delete(dt.addedRows, handle)
+	dt.deletedRows[handle] = struct{}{}
+}
+
+func (udb *dirtyDB) TruncateTable(tid int64) {
+	dt := udb.getDirtyTable(tid)
+	dt.addedRows = make(map[int64][]types.Datum)
+	dt.truncated = true
+}
+
+func (udb *dirtyDB) getDirtyTable(tid int64) *dirtyTable {
+	dt, ok := udb.tables[tid]
+	if !ok {
+		dt = &dirtyTable{
+			addedRows:   make(map[int64][]types.Datum),
+			deletedRows: make(map[int64]struct{}),
+		}
+		udb.tables[tid] = dt
+	}
+	return dt
+}
+
+type dirtyTable struct {
+	// key is handle.
+	addedRows   map[int64][]types.Datum
+	deletedRows map[int64]struct{}
+	truncated   bool
+}
+
+type dirtyDBKeyType int
+
+func (u dirtyDBKeyType) String() string {
+	return "dirtyDBKeyType"
+}
+
+// DirtyDBKey is the key to *dirtyDB for a context.
+const DirtyDBKey dirtyDBKeyType = 1
+
+func getDirtyDB(ctx context.Context) *dirtyDB {
+	var udb *dirtyDB
+	x := ctx.Value(DirtyDBKey)
+	if x == nil {
+		udb = &dirtyDB{tables: make(map[int64]*dirtyTable)}
+		ctx.SetValue(DirtyDBKey, udb)
+	} else {
+		udb = x.(*dirtyDB)
+	}
+	return udb
+}
+
+// UnionScanExec merges the rows from dirty table and the rows from XAPI request.
+type UnionScanExec struct {
```

Comment:
Yes, but all the executors are public right now, so we better keep it consistent.